spark3-extensions/src/main/scala/org/apache/spark/sql/catalyst/optimizer/RewriteMergeInto.scala

```diff
@@ -58,53 +61,136 @@ case class RewriteMergeInto(conf: SQLConf) extends Rule[LogicalPlan] with Rewrit
 
   override def apply(plan: LogicalPlan): LogicalPlan = {
     plan resolveOperators {
+      case MergeIntoTable(target: DataSourceV2Relation, source: LogicalPlan, cond, matchedActions, notMatchedActions)
+          if matchedActions.isEmpty =>
+
+        val (_, targetTableScan) = buildTargetScan(target, source, cond)
```

Comment:
@aokolnychyi, @dilipbiswal: in this case, should we still build the dynamically pruned target scan? It seems like that is just doing the join work twice. I'll work on removing it if you agree.

Reply:
Actually, this was fairly easy so I just implemented it. I can revert it if you think it is better with the dynamic filtering.

Reply:
@rdblue Currently the dynamically pruned target scan does the cardinality check Ryan. I haven't looked at the pr in detail. But just wanted to point out ..

Reply:
Do we need to perform the check if any match results in discarding the source row? I guess we could turn off the optimization if the cardinality check is enabled.

Reply:
+1 Ryan.

Reply:
Like discussed, no need to run the cardinality check for this branch.